File: rules.js
// Ki Shogi - Core rule engine
const { BLACK, WHITE, CUBE_GYOKU, CUBE_KI, FACE_CUBE, MOVES, OPPOSITES, PROMOTIONS, CUBE_FACES, resolveDirs } = require('./types');
const { MODE } = require('./state');

// Chebyshev distance
function chebDist(x1, y1, x2, y2) {
  return Math.max(Math.abs(x1 - x2), Math.abs(y1 - y2));
}

// Is position adjacent to any Gyoku or supported Ki on the board?
function isAdjacentToAnyGyoku(state, x, y) {
  for (const p of state.pieces) {
    if (p.cube === CUBE_GYOKU && p.onBoard && chebDist(x, y, p.x, p.y) === 1) return true;
  }
  // Check adjacency to supported Ki cubes
  for (const ki of getSupportedKi(state)) {
    if (chebDist(x, y, ki.x, ki.y) === 1) return true;
  }
  return false;
}

// BFS: find all Ki cubes connected to a Gyoku via adjacent Ki chain
function getSupportedKi(state) {
  const kiPieces = state.pieces.filter(p => p.cube === CUBE_KI && p.onBoard);
  if (kiPieces.length === 0) return [];
  const supported = new Set();
  const queue = [];
  for (const ki of kiPieces) {
    for (const g of state.pieces) {
      if (g.cube === CUBE_GYOKU && g.onBoard && chebDist(ki.x, ki.y, g.x, g.y) <= 2) {
        supported.add(ki.id);
        queue.push(ki);
        break;
      }
    }
  }
  while (queue.length > 0) {
    const cur = queue.shift();
    for (const ki of kiPieces) {
      if (!supported.has(ki.id) && chebDist(ki.x, ki.y, cur.x, cur.y) === 1) {
        supported.add(ki.id);
        queue.push(ki);
      }
    }
  }
  return kiPieces.filter(p => supported.has(p.id));
}

// Is position supported for a given cube type?
function isPieceSupported(state, x, y, cube) {
  if (cube === CUBE_KI) {
    for (const p of state.pieces) {
      if (p.cube === CUBE_GYOKU && p.onBoard && chebDist(x, y, p.x, p.y) <= 2) return true;
    }
    for (const ki of getSupportedKi(state)) {
      if (chebDist(x, y, ki.x, ki.y) === 1) return true;
    }
    return false;
  }
  return isAdjacentToAnyGyoku(state, x, y);
}

// Is position adjacent to a specific Gyoku?
function isAdjacentToGyoku(gyoku, x, y) {
  return gyoku.onBoard && chebDist(x, y, gyoku.x, gyoku.y) === 1;
}

// Get all direction vectors for a face+owner
function getMoveVectors(face, owner) {
  const moveDef = MOVES[face];
  if (!moveDef) return [];
  const result = [];
  for (const comp of moveDef) {
    for (const cat of comp.dirs) {
      for (const [dx, dy] of resolveDirs(cat, owner)) {
        result.push({ dx, dy, slide: comp.mode === 'slide' });
      }
    }
  }
  return result;
}

// Generate legal moves for a piece on the board
function getLegalMoves(state, piece) {
  if (!piece.onBoard) return [];
  const boardMap = state.buildBoardMap();
  const vectors = getMoveVectors(piece.face, piece.owner);
  const moves = [];
  const opponent = piece.owner === BLACK ? WHITE : BLACK;

  for (const { dx, dy, slide } of vectors) {
    const maxDist = slide ? 20 : 1;
    for (let d = 1; d <= maxDist; d++) {
      const nx = piece.x + dx * d;
      const ny = piece.y + dy * d;
      const key = `${nx},${ny}`;
      const target = boardMap.get(key);

      if (target) {
        // Can capture opponent's non-Gyoku piece
        if (target.owner === opponent) {
          // Gyoku special: can capture if it's not the opponent Gyoku... wait
          // Actually Gyoku CAN be captured? No - "The Gyoku cannot be left on an attacked position"
          // So we can move to capture any opponent piece
          if (wouldBeLegal(state, piece, nx, ny, target)) {
            moves.push({ x: nx, y: ny, capture: target });
          }
        }
        break; // blocked
      }

      // Empty square - check adjacency and legality
      if (wouldBeLegal(state, piece, nx, ny, null)) {
        moves.push({ x: nx, y: ny, capture: null });
      }
    }
  }
  return moves;
}

// Check if moving piece to (tx,ty) capturing target would be legal
function wouldBeLegal(state, piece, tx, ty, capturedPiece) {
  // Simulate the move
  const origX = piece.x, origY = piece.y;
  const origFace = piece.face;
  piece.x = tx; piece.y = ty;

  if (capturedPiece) {
    capturedPiece.x = null; capturedPiece.y = null;
    // Promotion face doesn't matter for legality check
  } else {
    // Flip to opposite
    if (piece.cube !== CUBE_GYOKU && OPPOSITES[piece.face]) {
      piece.face = OPPOSITES[piece.face];
    }
  }

  let legal = true;

  // Gyoku distance constraint
  if (piece.cube === CUBE_GYOKU) {
    const otherGyoku = state.pieces.find(p => p.cube === CUBE_GYOKU && p.id !== piece.id && p.onBoard);
    if (otherGyoku && chebDist(tx, ty, otherGyoku.x, otherGyoku.y) !== 2) {
      legal = false;
    }
  }

  // Piece must land adjacent to some Gyoku (unless it IS a Gyoku, or capturing Gyoku/Ki cube)
  if (legal && piece.cube !== CUBE_GYOKU && !(capturedPiece && (capturedPiece.cube === CUBE_GYOKU || capturedPiece.cube === CUBE_KI))) {
    if (!isPieceSupported(state, tx, ty, piece.cube)) legal = false;
  }

  // Own Gyoku must not be in check after move
  if (legal) {
    if (isInCheck(state, piece.owner)) legal = false;
  }

  // Restore
  piece.x = origX; piece.y = origY; piece.face = origFace;
  if (capturedPiece) {
    capturedPiece.x = tx; capturedPiece.y = ty;
  }

  return legal;
}

// Is the given player's Gyoku under attack?
function isInCheck(state, owner) {
  const gyoku = state.getGyoku(owner);
  if (!gyoku || !gyoku.onBoard) return false;
  const opponent = owner === BLACK ? WHITE : BLACK;

  for (const p of state.pieces) {
    if (p.owner !== opponent || !p.onBoard) continue;
    const vectors = getMoveVectors(p.face, p.owner);
    const boardMap = state.buildBoardMap();
    for (const { dx, dy, slide } of vectors) {
      const maxDist = slide ? 20 : 1;
      for (let d = 1; d <= maxDist; d++) {
        const nx = p.x + dx * d;
        const ny = p.y + dy * d;
        if (nx === gyoku.x && ny === gyoku.y) return true;
        if (boardMap.has(`${nx},${ny}`)) break;
      }
    }
  }
  return false;
}

// Get legal drop positions for a piece with a given face
function getLegalDrops(state, owner, face) {
  const gyoku = state.getGyoku(owner);
  if (!gyoku || !gyoku.onBoard) return [];
  const opponent = owner === BLACK ? WHITE : BLACK;
  const otherGyoku = state.getGyoku(opponent);
  const drops = [];

  // Ki cube can drop at distance 1 or 2 from own Gyoku; others only distance 1
  const isKi = FACE_CUBE[face] === CUBE_KI;
  const range = isKi ? 2 : 1;

  for (let dx = -range; dx <= range; dx++) {
    for (let dy = -range; dy <= range; dy++) {
      if (dx === 0 && dy === 0) continue;
      const dist = Math.max(Math.abs(dx), Math.abs(dy));
      if (dist > range) continue;
      const nx = gyoku.x + dx;
      const ny = gyoku.y + dy;

      if (state.getPieceAt(nx, ny)) continue;
      if (otherGyoku && otherGyoku.onBoard && chebDist(nx, ny, otherGyoku.x, otherGyoku.y) <= 1) continue;

      if (wouldDropCauseCheck(state, owner, face, nx, ny)) continue;

      // Drop must resolve own check if in check
      const tempPiece = { owner, face, onBoard: true, x: nx, y: ny, cube: FACE_CUBE[face] || 'temp', id: -2, key: `${nx},${ny}` };
      state.pieces.push(tempPiece);
      const stillInCheck = isInCheck(state, owner);
      state.pieces.pop();
      if (stillInCheck) continue;

      drops.push({ x: nx, y: ny });
    }
  }
  return drops;
}

function wouldDropCauseCheck(state, owner, face, x, y) {
  const opponent = owner === BLACK ? WHITE : BLACK;
  if (isInCheck(state, opponent)) return false; // already in check, drop didn't cause it
  const tempPiece = { owner, face, onBoard: true, x, y, cube: 'temp', id: -1, key: `${x},${y}` };
  state.pieces.push(tempPiece);
  const check = isInCheck(state, opponent);
  state.pieces.pop();
  return check;
}

// Find stranded pieces after a move and capture them
function handleStranding(state) {
  const captured = [];
  for (const p of state.pieces) {
    if (!p.onBoard || p.cube === CUBE_GYOKU || p.owner !== state.turn) continue;
    if (!isPieceSupported(state, p.x, p.y, p.cube)) {
      captured.push(p);
    }
  }
  const opponent = state.turn === BLACK ? WHITE : BLACK;
  for (const p of captured) {
    // Stranded pieces captured by opponent of the piece's owner
    const captor = p.owner === state.turn ? opponent : state.turn;
    p.x = null; p.y = null;
    p.owner = captor;
    // Reset to default face of cube
    const defaults = { hi: 'Hi', kaku: 'Kaku', ki: 'Ki' };
    if (defaults[p.cube]) p.face = defaults[p.cube];
  }
  return captured;
}

// Check if player has any legal action (move or drop)
function hasLegalAction(state, owner) {
  // Check board moves
  for (const p of state.pieces) {
    if (p.owner === owner && p.onBoard) {
      if (getLegalMoves(state, p).length > 0) return true;
    }
  }
  // Check drops
  const hand = state.getHandPieces(owner);
  for (const p of hand) {
    const faces = p.cube === CUBE_GYOKU ? ['Gyoku'] : require('./types').CUBE_FACES[p.cube];
    for (const face of faces) {
      if (getLegalDrops(state, owner, face).length > 0) return true;
    }
  }
  return false;
}

// Valid positions for White Gyoku placement (Chebyshev distance exactly 2 from Black Gyoku)
function getWhiteGyokuPositions(state) {
  const bg = state.getGyoku(BLACK);
  if (!bg || !bg.onBoard) return [];
  const positions = [];
  for (let dx = -2; dx <= 2; dx++) {
    for (let dy = -2; dy <= 2; dy++) {
      if (Math.max(Math.abs(dx), Math.abs(dy)) === 2) {
        positions.push({ x: bg.x + dx, y: bg.y + dy });
      }
    }
  }
  return positions;
}

module.exports = {
  chebDist, isAdjacentToAnyGyoku, isAdjacentToGyoku, isPieceSupported,
  getLegalMoves, getLegalDrops, isInCheck, handleStranding,
  hasLegalAction, getWhiteGyokuPositions, getMoveVectors,
};

File: ki-shogi-rs/src/rules.rs
// Ki Shogi - Core rule engine

use crate::types::{Owner, Cube, Face, get_move_vectors};
use crate::state::{GameState, LegalMove, Piece};
use std::collections::HashSet;

/// Chebyshev distance
pub fn cheb_dist(x1: i32, y1: i32, x2: i32, y2: i32) -> i32 {
    (x1 - x2).abs().max((y1 - y2).abs())
}

/// BFS: find all Ki cubes connected to a Gyoku via adjacent Ki chain
fn get_supported_ki(state: &GameState) -> HashSet<usize> {
    let ki_pieces: Vec<(usize, i32, i32)> = state.pieces.iter().enumerate()
        .filter(|(_, p)| p.cube == Cube::Ki && p.on_board())
        .map(|(i, p)| (i, p.x.unwrap(), p.y.unwrap()))
        .collect();
    if ki_pieces.is_empty() { return HashSet::new(); }

    let mut supported = HashSet::new();
    let mut queue = Vec::new();
    for &(idx, kx, ky) in &ki_pieces {
        for p in &state.pieces {
            if p.cube == Cube::Gyoku && p.on_board() {
                if cheb_dist(kx, ky, p.x.unwrap(), p.y.unwrap()) <= 2 {
                    supported.insert(idx);
                    queue.push((idx, kx, ky));
                    break;
                }
            }
        }
    }
    while let Some((_, cx, cy)) = queue.pop() {
        for &(idx, kx, ky) in &ki_pieces {
            if !supported.contains(&idx) && cheb_dist(kx, ky, cx, cy) == 1 {
                supported.insert(idx);
                queue.push((idx, kx, ky));
            }
        }
    }
    supported
}

/// Is position supported for a given cube type?
pub fn is_piece_supported(state: &GameState, x: i32, y: i32, cube: Cube) -> bool {
    let supported_ki = get_supported_ki(state);
    if cube == Cube::Ki {
        for p in &state.pieces {
            if p.cube == Cube::Gyoku && p.on_board()
                && cheb_dist(x, y, p.x.unwrap(), p.y.unwrap()) <= 2
            { return true; }
        }
        for &idx in &supported_ki {
            let ki = &state.pieces[idx];
            if cheb_dist(x, y, ki.x.unwrap(), ki.y.unwrap()) == 1 { return true; }
        }
        return false;
    }
    // Normal piece: adjacent to any Gyoku or supported Ki (dist 1)
    for p in &state.pieces {
        if p.cube == Cube::Gyoku && p.on_board()
            && cheb_dist(x, y, p.x.unwrap(), p.y.unwrap()) == 1
        { return true; }
    }
    for &idx in &supported_ki {
        let ki = &state.pieces[idx];
        if cheb_dist(x, y, ki.x.unwrap(), ki.y.unwrap()) == 1 { return true; }
    }
    false
}

/// Is the given player's Gyoku under attack?
pub fn is_in_check(state: &GameState, owner: Owner) -> bool {
    let gi = match state.gyoku(owner) {
        Some(i) => i,
        None => return false,
    };
    let g = &state.pieces[gi];
    if !g.on_board() { return false; }
    let (gx, gy) = (g.x.unwrap(), g.y.unwrap());
    let opponent = owner.opponent();
    let board_map = state.build_board_map();

    for p in &state.pieces {
        if p.owner != opponent || !p.on_board() { continue; }
        let (px, py) = (p.x.unwrap(), p.y.unwrap());
        for (dx, dy, is_slide) in get_move_vectors(p.face, p.owner) {
            let max_d = if is_slide { 20 } else { 1 };
            for d in 1..=max_d {
                let nx = px + dx * d;
                let ny = py + dy * d;
                if nx == gx && ny == gy { return true; }
                if board_map.contains_key(&format!("{},{}", nx, ny)) { break; }
            }
        }
    }
    false
}

/// Check if moving piece to (tx,ty) would be legal. Mutates and restores state.
fn would_be_legal(state: &mut GameState, pidx: usize, tx: i32, ty: i32, cap_idx: Option<usize>) -> bool {
    let orig_x = state.pieces[pidx].x;
    let orig_y = state.pieces[pidx].y;
    let orig_face = state.pieces[pidx].face;
    let p_cube = state.pieces[pidx].cube;
    let p_owner = state.pieces[pidx].owner;
    let p_id = state.pieces[pidx].id;

    state.pieces[pidx].x = Some(tx);
    state.pieces[pidx].y = Some(ty);

    let mut cap_ox = None;
    let mut cap_oy = None;
    if let Some(ci) = cap_idx {
        cap_ox = state.pieces[ci].x;
        cap_oy = state.pieces[ci].y;
        state.pieces[ci].x = None;
        state.pieces[ci].y = None;
    } else if p_cube != Cube::Gyoku {
        if let Some(opp) = state.pieces[pidx].face.opposite() {
            state.pieces[pidx].face = opp;
        }
    }

    let mut legal = true;

    if p_cube == Cube::Gyoku {
        for i in 0..state.pieces.len() {
            let p = &state.pieces[i];
            if p.cube == Cube::Gyoku && p.id != p_id && p.on_board() {
                if cheb_dist(tx, ty, p.x.unwrap(), p.y.unwrap()) != 2 {
                    legal = false;
                }
                break;
            }
        }
    }

    if legal && p_cube != Cube::Gyoku {
        let skip = cap_idx.map_or(false, |ci| {
            matches!(state.pieces[ci].cube, Cube::Gyoku | Cube::Ki)
        });
        if !skip && !is_piece_supported(state, tx, ty, p_cube) {
            legal = false;
        }
    }

    if legal && is_in_check(state, p_owner) {
        legal = false;
    }

    state.pieces[pidx].x = orig_x;
    state.pieces[pidx].y = orig_y;
    state.pieces[pidx].face = orig_face;
    if let Some(ci) = cap_idx {
        state.pieces[ci].x = cap_ox;
        state.pieces[ci].y = cap_oy;
    }
    legal
}

/// Generate legal moves for a piece on the board
pub fn get_legal_moves(state: &mut GameState, pidx: usize) -> Vec<LegalMove> {
    if !state.pieces[pidx].on_board() { return Vec::new(); }
    let face = state.pieces[pidx].face;
    let owner = state.pieces[pidx].owner;
    let px = state.pieces[pidx].x.unwrap();
    let py = state.pieces[pidx].y.unwrap();
    let opponent = owner.opponent();
    let board_map = state.build_board_map();
    let vectors = get_move_vectors(face, owner);
    let mut moves = Vec::new();

    for (dx, dy, is_slide) in vectors {
        let max_d = if is_slide { 20 } else { 1 };
        for d in 1..=max_d {
            let nx = px + dx * d;
            let ny = py + dy * d;
            let key = format!("{},{}", nx, ny);
            if let Some(&ti) = board_map.get(&key) {
                if state.pieces[ti].owner == opponent {
                    if would_be_legal(state, pidx, nx, ny, Some(ti)) {
                        moves.push(LegalMove { x: nx, y: ny, capture: Some(ti) });
                    }
                }
                break;
            }
            if would_be_legal(state, pidx, nx, ny, None) {
                moves.push(LegalMove { x: nx, y: ny, capture: None });
            }
        }
    }
    moves
}

/// Would dropping a piece with given face at (x,y) give check to opponent?
fn would_drop_cause_check(state: &mut GameState, owner: Owner, face: Face, x: i32, y: i32) -> bool {
    let opponent = owner.opponent();
    if is_in_check(state, opponent) { return false; }
    let temp = Piece::new(9999, owner, face.cube(), face);
    let idx = state.pieces.len();
    state.pieces.push(temp);
    state.pieces[idx].x = Some(x);
    state.pieces[idx].y = Some(y);
    let check = is_in_check(state, opponent);
    state.pieces.pop();
    check
}

/// Get legal drop positions for a piece with a given face
pub fn get_legal_drops(state: &mut GameState, owner: Owner, face: Face) -> Vec<(i32, i32)> {
    let gi = match state.gyoku(owner) { Some(i) => i, None => return Vec::new() };
    if !state.pieces[gi].on_board() { return Vec::new(); }
    let (gx, gy) = (state.pieces[gi].x.unwrap(), state.pieces[gi].y.unwrap());
    let opponent = owner.opponent();
    let ogi = state.gyoku(opponent);

    let is_ki = face.cube() == Cube::Ki;
    let range: i32 = if is_ki { 2 } else { 1 };
    let mut drops = Vec::new();

    for dx in -range..=range {
        for dy in -range..=range {
            if dx == 0 && dy == 0 { continue; }
            if dx.abs().max(dy.abs()) > range { continue; }
            let nx = gx + dx;
            let ny = gy + dy;

            if state.piece_at(nx, ny).is_some() { continue; }

            if let Some(oi) = ogi {
                let og = &state.pieces[oi];
                if og.on_board() && cheb_dist(nx, ny, og.x.unwrap(), og.y.unwrap()) <= 1 {
                    continue;
                }
            }

            if would_drop_cause_check(state, owner, face, nx, ny) { continue; }

            // Drop must resolve own check
            let temp = Piece::new(9998, owner, face.cube(), face);
            let tidx = state.pieces.len();
            state.pieces.push(temp);
            state.pieces[tidx].x = Some(nx);
            state.pieces[tidx].y = Some(ny);
            let still = is_in_check(state, owner);
            state.pieces.pop();
            if still { continue; }

            drops.push((nx, ny));
        }
    }
    drops
}

/// Find stranded pieces and capture them for opponent
pub fn handle_stranding(state: &mut GameState) {
    let turn = state.turn;
    let opponent = turn.opponent();
    let stranded: Vec<usize> = state.pieces.iter().enumerate()
        .filter(|(_, p)| {
            p.on_board() && p.cube != Cube::Gyoku && p.owner == turn
                && !is_piece_supported(state, p.x.unwrap(), p.y.unwrap(), p.cube)
        })
        .map(|(i, _)| i)
        .collect();

    for idx in stranded {
        state.pieces[idx].x = None;
        state.pieces[idx].y = None;
        state.pieces[idx].owner = opponent;
        state.pieces[idx].face = match state.pieces[idx].cube {
            Cube::Hi => Face::Hi,
            Cube::Kaku => Face::Kaku,
            Cube::Ki => Face::Ki,
            _ => state.pieces[idx].face,
        };
    }
}

/// Check if player has any legal action (move or drop)
pub fn has_legal_action(state: &mut GameState, owner: Owner) -> bool {
    let board_idxs: Vec<usize> = state.pieces.iter().enumerate()
        .filter(|(_, p)| p.owner == owner && p.on_board())
        .map(|(i, _)| i).collect();
    for pidx in board_idxs {
        if !get_legal_moves(state, pidx).is_empty() { return true; }
    }
    let hand_idxs = state.hand_pieces_for(owner);
    for &pi in &hand_idxs {
        let cube = state.pieces[pi].cube;
        for &face in Face::cube_faces(cube) {
            if !get_legal_drops(state, owner, face).is_empty() { return true; }
        }
    }
    false
}

/// Valid positions for White Gyoku (Chebyshev distance exactly 2 from Black Gyoku)
pub fn get_white_gyoku_positions(state: &GameState) -> Vec<(i32, i32)> {
    let gi = match state.gyoku(Owner::Black) { Some(i) => i, None => return Vec::new() };
    let g = &state.pieces[gi];
    if !g.on_board() { return Vec::new(); }
    let (bx, by) = (g.x.unwrap(), g.y.unwrap());
    let mut pos = Vec::new();
    for dx in -2..=2i32 {
        for dy in -2..=2i32 {
            if dx.abs().max(dy.abs()) == 2 {
                pos.push((bx + dx, by + dy));
            }
        }
    }
    pos
}
